proposals/0279-multiple-trailing-closures.md

```diff
@@ -1,422 +1,369 @@
-many existing APIs, without requiring the API authors to make changes.
+This introduces a zero-lookahead ambiguity between the start of a labeled trailing closure and the start of either a new expression, a labeled statement, or a `default:` label followed by `'{'`.  The first two ambiguities can be resolved by looking forward at most two tokens, because (`identifier|keyword|'_') ':'` can never start an expression, only a labeled statement, and the statement in a labeled statement can never start with `'{'` while `expr-closure` must start with it.  The ambiguity with `default:` can be resolved by not allowing the unescaped `default` keyword as a label in this syntax; it can still be used if necessary by escaping it (i.e. \``default`\``:`).  Source compatibility requires the existing use of `default:` to be preferred, and it's better to do this uniformly (even if the syntax does not appear in a `switch`) in order to discourage the use of `default` as a trailing-closure label in APIs, rather than leaving a trap in the language if an API like that is used within a `switch`.
+
+The labeled trailing closures are associated with the base expression in the same way as the unlabeled trailing closure is today:
 
-## Detailed design
+* if the base expression is a call, they are added as extra arguments to that call;
+* if the base expression is a subscript, they are added as extra index arguments to that subscript,
+* otherwise, an implicit call of the base expression is created using only the trailing closures as arguments.
 
-Changes required to implement new syntax are constrained solely to the parser, namely to the parsing of call arguments.
+The existing trailing-closure feature requires special treatment of the trailing-closure argument by the type checker because of the special power of label omission: the trailing closure can be passed to a parameter that would ordinarily require an argument label.  Currently, the special treatment is specific to the final argument.  Because this proposal still has an unlabeled trailing-closure argument, we have to generalize that treatment to allow label omission at an intermediate argument.
 
-It's possible to do an early syntax transformation which would consider each block to be a regular argument and adjust locations of parentheses if necessary.
-For type checker perspective this means that calls with new syntax would just be regular calls which require no special handling:
+Note that labeled trailing closures are required to match labels with a parameter.  A labeled trailing closure can use the special label `_` to indicate that it matches an unlabeled parameter, but this *only* matches an unlabeled parameter; it does not have the special label-omission powers of the initial unlabeled trailing closure.  For example:
 
 ```swift
-when(2 < 3) {
-  then: { ... }
-  else: { ... }
+func pointFromClosures(x: () -> Int, _ y: () -> Int) -> (Int, Int) { (x,y) }
```

Comment:
* Call `x()` and `y()` in body
* Reformat as it's now longer than 80 characters

```suggestion
func pointFromClosures(
  x: () -> Int,
  _ y: () -> Int
) -> (Int, Int) {
  (x(), y())
}
```